packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts

```diff
@@ -0,0 +1,102 @@
+import { useCallback, useMemo, useState } from 'react'
+import stringify from 'json-stable-stringify'
+import { pick } from 'lodash-es'
+import type { TransactionConfig } from 'web3-core'
+import {
+    ChainId,
+    TransactionState,
+    TransactionStateType,
+    useAccount,
+    useChainId,
+    useGasPrice,
+    useNonce,
+    useWeb3,
+} from '@masknet/web3-shared'
+import type { TradeComputed, SwapQuoteOneResponse } from '../../types'
+
+export function useTradeCallback(tradeComputed: TradeComputed<SwapQuoteOneResponse> | null) {
+    const web3 = useWeb3()
+    const gasPrice = useGasPrice()
+    const account = useAccount()
+    const nonce = useNonce(account)
+    const chainId = useChainId()
+    const [tradeState, setTradeState] = useState<TransactionState>({
+        type: TransactionStateType.UNKNOWN,
+    })
+
+    // compose transaction config
+    const config = useMemo(() => {
+        if (!account || !tradeComputed?.trade_ || ![ChainId.Mainnet, ChainId.BSC, ChainId.Matic].includes(chainId))
+            return null
+        return {
+            from: account,
+            ...pick(tradeComputed.trade_, ['to', 'data', 'value', 'gas', 'gasPrice']),
+        } as TransactionConfig
+    }, [account, tradeComputed])
+
+    const tradeCallback = useCallback(async () => {
+        // validate config
+        if (!account || !config) {
+            setTradeState({
+                type: TransactionStateType.UNKNOWN,
+            })
+            return
+        }
+
+        // start waiting for provider to confirm tx
+        setTradeState({
+            type: TransactionStateType.WAIT_FOR_CONFIRMING,
+        })
+
+        // estimate transaction
+        try {
+            await web3.eth.estimateGas(config)
+        } catch {
+            // for some transactions will always fail if we do estimation before a kick to the chain
```

Comment:
Please don't copy this here. Instead, it should implement its error handler.

Reply:
Ok @guanbinrui, I will implement this.